File: JavaScript/1155-Number of Dice Rolls With Target Sum.js
/**
 * @param {number} d
 * @param {number} f
 * @param {number} target
 * @return {number}
 */
var numRollsToTarget = function(d, f, target) {
    const kMod = 1000000007;
    let memo = new Map();
    let dfs = (dice, targetRemain) =>{
        if (dice === 0){
            if (targetRemain === 0) return 1;
            return 0;
        }
        
        if (dice*f < targetRemain || targetRemain < dice){
            return 0;
        }
        
        if (memo.has(`${dice}-${targetRemain}`) ){
            return memo.get(`${dice}-${targetRemain}`);
        }
        
        let res = 0;
        for (let nxt = 1; nxt <= f; nxt++){
            if (nxt > targetRemain) break;
            res = (res + dfs(dice-1, targetRemain-nxt))%kMod;
        }
        memo.set(`${dice}-${targetRemain}`, res)
        return res;
    }
    return dfs(d, target);
};

File: C++/1155-Number of Dice Rolls With Target Sum.cpp
class Solution {
    int kMod = 1e9+7;
public:
    int dfs(vector<vector<int>>& memo, int n, int k, int remain){
        if (memo[n][remain] != -1) return memo[n][remain];
        if (n == 1){
            if (remain <= k && remain >= 1) return 1;
            else return 0;
        }
        
        int res = 0;
        for (int nxt=1; nxt <= k; nxt++){
            if (remain-nxt < n-1) break;
            res = (res + dfs(memo, n-1, k, remain-nxt))%kMod;
        }
        memo[n][remain] = res;
        return res;
    }
    
    int numRollsToTarget(int n, int k, int target) {
        vector<vector<int>> memo(n+1, vector<int>(target+1, -1));
        return dfs(memo, n, k, target);
    }
};